src/reduced_basis/rb_construction_base.C

```diff
@@ -469,115 +505,67 @@ void RBConstructionBase<Base>::generate_training_parameters_random(const Paralle
     }
 
   // initialize training_parameters_in
-    for (const auto & pr : min_parameters)
-      {
-        const std::string & param_name = pr.first;
-        training_parameters_in[param_name] = NumericVector<Number>::build(communicator);
-
-        if (!serial_training_set)
-          {
-            // Calculate the number of training parameters local to this processor
-            unsigned int n_local_training_samples;
-            unsigned int quotient  = n_training_samples_in/communicator.size();
-            unsigned int remainder = n_training_samples_in%communicator.size();
-            if (communicator.rank() < remainder)
-              n_local_training_samples = (quotient + 1);
-            else
-              n_local_training_samples = quotient;
-
-            training_parameters_in[param_name]->init(n_training_samples_in, n_local_training_samples, false, PARALLEL);
-          }
-        else
-          {
-            training_parameters_in[param_name]->init(n_training_samples_in, false, SERIAL);
-          }
-      }
+  const auto &[n_local_training_samples, first_local_index] =
+      calculate_n_local_samples_and_index(communicator, n_global_training_samples_in,
+                                          serial_training_set);
+  for (const auto & it : min_parameters)
+    local_training_parameters_in[it.first] = std::vector<Real>(n_local_training_samples);
 
   // finally, set the values
-  for (auto & pr : training_parameters_in)
+  for (auto & [param_name, sample_vector]: local_training_parameters_in)
     {
-      const std::string & param_name = pr.first;
-      NumericVector<Number> * training_vector = pr.second.get();
-
-      numeric_index_type first_index = training_vector->first_local_index();
-      for (auto i : make_range(training_vector->local_size()))
+      for (auto i : make_range(n_local_training_samples))
         {
-          numeric_index_type index = first_index + i;
           Real random_number = static_cast<Real>(std::rand()) / RAND_MAX; // in range [0,1]
 
           // Generate log10 scaled training parameters
-          if (log_param_scale[param_name])
+          if (log_param_scale.at(param_name))
```

Comment:
Use `libmesh_map_find(log_param_scale, param_name)` instead of `at` here. The helper function gives a better error message that includes a stack trace when `param_name` is not in the map.

Reply:
Sure. I switched to `.at()` from `[]` when I made the `log_param_scale` const, not for any explicit need of bounds checking.